### PR TITLE
Create controller for the Loans resource

### PR DESCRIPTION
**Pull request by mreis-ch**
Create controller for Loan resource
Create LoansToDirectors interceptor
Update methods that create self links in LoanServiceImpl
Add tests for controller and interceptor
Update test URI for LoanServiceImplTest

BI-4039

### CODE REVIEW

File: src/test/java/uk/gov/companieshouse/api/accounts/controller/LoansControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+
+        verify(loanService, times(1))
+                .update(loanRest, transaction, COMPANY_ACCOUNT_ID, request);
+        verify(apiResponseMapper, times(1))
+                .map(responseObject.getStatus(), responseObject.getData(), responseObject.getErrors());
+    }
+
+    @Test
+    @DisplayName("Tests the update of a Loan when the Loan ID doesnt exist")
+    void UpdateDirectorResourceWhenIdIsNull() throws DataException {
```
Comment (hwilliams1) [inline src/test/java/uk/gov/companieshouse/api/accounts/controller/LoansControllerTest.java:247]: method name should start with a lowercase letter -updateDirectorResourceWhenIdIsNull

File: src/main/java/uk/gov/companieshouse/api/accounts/interceptor/LoansToDirectorsInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            String smallFullLoansLink =
+                    smallFull.getLinks().get(SmallFullLinkType.LOANS_TO_DIRECTORS.getLink());
+
+            LoansToDirectors loansToDirectors = responseObject.getData();
+            String loansToDirectorsSelfLink = loansToDirectors.getLinks().get(BasicLinkType.SELF.getLink());
+
+            if (!loansToDirectorsSelfLink.equals(smallFullLoansLink)) {
+
+                LoggingHelper.logInfo(
+                        companyAccountId, transaction, "Loans to directors link not present in small full resource", request);
```
Comment (hwilliams1) [inline src/main/java/uk/gov/companieshouse/api/accounts/interceptor/LoansToDirectorsInterceptor.java:71]: We shouldn't be logging objects, we should be logging the required fields in the objects instead

File: src/main/java/uk/gov/companieshouse/api/accounts/interceptor/LoansToDirectorsInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                return false;
+            }
+
+            request.setAttribute(AttributeName.LOANS_TO_DIRECTORS.getValue(), loansToDirectors);
+            return true;
+
+        } catch (DataException e) {
+
+            LoggingHelper.logException(
+                    companyAccountId, transaction, "Failed to retrieve loans to directors resource", e, request);
```
Comment (hwilliams1) [inline src/main/java/uk/gov/companieshouse/api/accounts/interceptor/LoansToDirectorsInterceptor.java:83]: We shouldn't be logging objects, we should log the required fields from it

File: src/main/java/uk/gov/companieshouse/api/accounts/interceptor/LoansToDirectorsInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+        String companyAccountId = pathVariables.get(COMPANY_ACCOUNTS_ID_PATH_VARIABLE);
+
+        try {
+            ResponseObject<LoansToDirectors> responseObject = loansToDirectorsService.find(companyAccountId, request);
+
+            if (!responseObject.getStatus().equals(ResponseStatus.FOUND)) {
+
+                LoggingHelper.logInfo(
+                        companyAccountId, transaction, "Loans to directors not found", request);
```
Comment (hwilliams1) [inline src/main/java/uk/gov/companieshouse/api/accounts/interceptor/LoansToDirectorsInterceptor.java:53]: We shouldn't be logging objects we should log the required fields from it

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/LoansController.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+                                 HttpServletRequest request) {
+
+        Transaction transaction = (Transaction) request
+                .getAttribute(AttributeName.TRANSACTION.getValue());
+
+        try {
+            ResponseObject<Loan> response =
+                    loanService.delete(companyAccountId, request);
+            return apiResponseMapper
+                    .map(response.getStatus(), response.getData(), response.getErrors());
```
Comment (hwilliams1) [inline src/main/java/uk/gov/companieshouse/api/accounts/controller/LoansController.java:141]: Shouldn't we be logging that this is occurring e.g. deleting loanId?

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/LoansController.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+        Transaction transaction = (Transaction) request.getAttribute(AttributeName.TRANSACTION.getValue());
+
+        try {
+            ResponseObject<Loan> response =
+                    loanService.
+                            update(loan, transaction, companyAccountId, request);
+
+            return apiResponseMapper
+                    .map(response.getStatus(), response.getData(), response.getErrors());
```
Comment (hwilliams1) [inline src/main/java/uk/gov/companieshouse/api/accounts/controller/LoansController.java:120]: We should be adding logging to this

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/LoansController.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+        Transaction transaction = (Transaction) request
+                .getAttribute(AttributeName.TRANSACTION.getValue());
+
+        try {
+            ResponseObject<Loan> response = loanService
+                    .findAll(transaction, companyAccountId, request);
+
+            return apiResponseMapper.mapGetResponseForMultipleResources(response.getDataForMultipleResources(), request);
+
```
Comment (hwilliams1) [inline src/main/java/uk/gov/companieshouse/api/accounts/controller/LoansController.java:92]: we should be adding logging to this

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/LoansController.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                              HttpServletRequest request) {
+
+        Transaction transaction = (Transaction) request
+                .getAttribute(AttributeName.TRANSACTION.getValue());
+
+        try {
+            ResponseObject<Loan> response = loanService
+                    .find(companyAccountId, request);
+
+            return apiResponseMapper.mapGetResponse(response.getData(), request);
```
Comment (hwilliams1) [inline src/main/java/uk/gov/companieshouse/api/accounts/controller/LoansController.java:70]: Shouldn't this be logged?

File: src/main/java/uk/gov/companieshouse/api/accounts/controller/LoansController.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    		HttpServletRequest request) {
+
+        Transaction transaction = (Transaction) request.getAttribute(AttributeName.TRANSACTION.getValue());
+
+        try {
+            ResponseObject<Loan> response =
+                    loanService.create(loan, transaction, companyAccountId, request);
+
+            return apiResponseMapper
+                    .map(response.getStatus(), response.getData(), response.getErrors());
```
Comment (hwilliams1) [inline src/main/java/uk/gov/companieshouse/api/accounts/controller/LoansController.java:49]: Shouldn't this be logged?

Comment (BSaunders95) [reply src/main/java/uk/gov/companieshouse/api/accounts/controller/LoansController.java:49]: Everything is logged in an interceptor (request and response) so logging in the controller isn't necessarily needed, but that's a preference thing

File: src/test/java/uk/gov/companieshouse/api/accounts/controller/LoansControllerTest.java (hunk 1 above)
Comment (mreis-ch) [reply src/test/java/uk/gov/companieshouse/api/accounts/controller/LoansControllerTest.java]: Done 👍

**Pull request merged**